File: prototype/node/src/app/core/reports/listing-status-tally.ts
import { LISTING_STATUSES, type ListingStatus } from '../listings/listing-status.ts'

export type ListingStatusCount = { status: ListingStatus; count: number }

/**
 * Every listing status a seller owns, in lifecycle order, so a dashboard tile
 * stays in place on a day nothing sold.
 */
export function listingStatusTally(
  countsByStatus: Partial<Record<ListingStatus, number>>,
): readonly ListingStatusCount[] {
  return LISTING_STATUSES.map((status) => ({ status, count: countsByStatus[status] ?? 0 }))
}

File: prototype/rails/src/app/domain/reports/listing_status_tally.rb
require_relative "listing_status_count"

module Domain
  module Reports
    # Every listing status a seller owns, in lifecycle order, so the dashboard
    # keeps its tiles in place on a day nothing sold.
    module ListingStatusTally
      module_function

      def from(counts_by_status)
        ::Listing.statuses.keys.map do |status|
          ListingStatusCount.new(status: status, count: counts_by_status.fetch(status, 0))
        end
      end

      def total(tally)
        tally.sum(&:count)
      end
    end
  end
end
